### PR TITLE
[SPARK-7780][MLLIB] intercept in logisticregressionwith lbfgs should not be regularized

### PR DESCRIPTION
**Pull request by holdenk**
The intercept in Logistic Regression represents a prior on categories which should not be regularized. In MLlib, the regularization is handled through Updater, and the Updater penalizes all the components without excluding the intercept which resulting poor training accuracy with regularization.
The new implementation in ML framework handles this properly, and we should call the implementation in ML from MLlib since majority of users are still using MLlib api.
Note that both of them are doing feature scalings to improve the convergence, and the only difference is ML version doesn't regularize the intercept. As a result, when lambda is zero, they will converge to the same solution.

Previously partially reviewed at https://github.com/apache/spark/pull/6386#issuecomment-168781424 re-opening for @dbtsai to review.

### CODE REVIEW

**Comment by holdenk** [comment]
Also related is https://github.com/apache/spark/pull/6771 which avoids the round trip through DataFrames (although since it now works on Instances rather than LabeledPoints under the hood it got a bit more complicated) so its probably OK to have the round trip (but just to keep as an option).

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 1]
```diff
@@ -247,10 +247,30 @@ class LogisticRegression @Since("1.2.0") (
...
-  override protected def train(dataset: DataFrame): LogisticRegressionModel = {
-    // Extract columns from data.  If dataset is persisted, do not persist oldDataset.
+  private var optInitialWeights: Option[Vector] = None
+  /** @group setParam */
+  private[spark] def setInitialWeights(value: Vector): this.type = {
+    this.optInitialWeights = Some(value)
+    this
+  }
+
+  /** Validate the initial weights, return an Option, if not the expected size return None
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Move `Validate .... into a new line`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 2]
```diff
@@ -247,10 +247,30 @@ class LogisticRegression @Since("1.2.0") (
   @Since("1.5.0")
   override def getThresholds: Array[Double] = super.getThresholds
 
-  override protected def train(dataset: DataFrame): LogisticRegressionModel = {
-    // Extract columns from data.  If dataset is persisted, do not persist oldDataset.
+  private var optInitialWeights: Option[Vector] = None
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: We started to change our terminology from `weights` into `coefficients` after the implementation of weighted LoR. Please change it into `optInitialCoefficients`.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 3]
```diff
@@ -247,10 +247,30 @@ class LogisticRegression @Since("1.2.0") (
...
+  /** @group setParam */
+  private[spark] def setInitialWeights(value: Vector): this.type = {
+    this.optInitialWeights = Some(value)
+    this
+  }
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: How about we follow https://github.com/apache/spark/pull/8972 , and have the following code. We can create another seprate JIRA for moving `setInitialModel` to public with a sharedParam.

```scala
  private var initialModel: Option[LogisticRegressionModel] = None

  private def setInitialModel(model: LogisticRegressionModel): this.type = {
    ...
    ...
    this
  }
```

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 4]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+        val lr = new org.apache.spark.ml.classification.LogisticRegression()
+        lr.setRegParam(optimizer.getRegParam())
+        lr.setElasticNetParam(elasticNetParam)
+        if (userSuppliedWeights) {
+          val initialWeightsWithIntercept = if (addIntercept) {
+            appendBias(initialWeights)
+          } else {
+            initialWeights
+          }
+          lr.setInitialWeights(initialWeightsWithIntercept)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Here will be

```scala
lr.setInitialModel(new org.apache.spark.ml.classification.LogisticRegressionModel(uid, initialWeights, 1))
```

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 5]
```diff
@@ -247,10 +247,31 @@ class LogisticRegression @Since("1.2.0") (
...
+   * Validate the initial weights, return an Option, if not the expected size return None
+   * and log a warning.
+   */
+  private def validateWeights(vectorOpt: Option[Vector], numFeatures: Int): Option[Vector] = {
+    vectorOpt.flatMap(vec =>
+      if (vec.size == numFeatures) {
+        Some(vec)
+      } else {
+        logWarning(
+          s"""Initial weights provided (${vec})did not match the expected size ${numFeatures}""")
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: btw, why `s"""`, also change `weights` to coefficients

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 3 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: So we have setInitialWeights on StreamingLogisticRegressionWithSGD - would it be better to have it match StreamingLogisticRegressionWithSGD ?

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 6]
```diff
@@ -247,10 +247,31 @@ class LogisticRegression @Since("1.2.0") (
...
+  private[spark] def setInitialWeights(value: Vector): this.type = {
+    this.optInitialCoefficients = Some(value)
+    this
+  }
+
+  /**
+   * Validate the initial weights, return an Option, if not the expected size return None
+   * and log a warning.
+   */
+  private def validateWeights(vectorOpt: Option[Vector], numFeatures: Int): Option[Vector] = {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: validateCoefficients

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 7]
```diff
@@ -247,10 +247,31 @@ class LogisticRegression @Since("1.2.0") (
...
+        logWarning(
+          s"""Initial weights provided (${vec})did not match the expected size ${numFeatures}""")
+        None
+      })
+  }
+
+  override protected[spark] def train(dataset: DataFrame): LogisticRegressionModel = {
     val w = if ($(weightCol).isEmpty) lit(1.0) else col($(weightCol))
-    val instances: RDD[Instance] = dataset.select(col($(labelCol)), w, col($(featuresCol))).map {
+    val instances = dataset.select(col($(labelCol)), w, col($(featuresCol))).map {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: why this line is changed?

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 8]
```diff
@@ -322,10 +343,12 @@ class LogisticRegression @Since("1.2.0") (
       new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
     }
 
-    val initialCoefficientsWithIntercept =
-      Vectors.zeros(if ($(fitIntercept)) numFeatures + 1 else numFeatures)
+    val numFeaturesWithIntercept = if ($(fitIntercept)) numFeatures + 1 else numFeatures
+    val userSuppliedCoefficients = validateWeights(optInitialCoefficients, numFeaturesWithIntercept)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: let's handle it through setInitialModel, and have another PR to make it public. Thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 9]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
       new LogisticRegressionModel(weights, intercept, numFeatures, numOfLinearPredictor + 1)
     }
   }
+
+  /**
+   * Run the algorithm with the configured parameters on an input RDD
+   * of LabeledPoint entries starting from the initial weights provided.
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Replace `algorithm` by `Logistic Regression`, and remove `starting from the initial weights provided`.

Add a new line between `of LabeledPoint entries` and `If a known updater is used`.

Actually, in ml version, disabling feature scaling is supported now. So please call ml implementation in this case.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 10]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+   * If a known updater is used calls the ml implementation, to avoid
+   * applying a regularization penalty to the intercept, otherwise
+   * defaults to the mllib implementation. If more than two classes
+   * or feature scaling is disabled, always uses mllib implementation.
+   * If using ml implementation, uses ml code to generate initial weights.
+   */
+  override def run(input: RDD[LabeledPoint]): LogisticRegressionModel = {
+    run(input, generateInitialWeights(input), false)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: `run(input, generateInitialWeights(input), userSuppliedWeights = false)`

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 11]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+   * Run the algorithm with the configured parameters on an input RDD
+   * of LabeledPoint entries starting from the initial weights provided.
+   * If a known updater is used calls the ml implementation, to avoid
+   * applying a regularization penalty to the intercept, otherwise
+   * defaults to the mllib implementation. If more than two classes
+   * or feature scaling is disabled, always uses mllib implementation.
+   * Uses user provided weights.
+   */
+  override def run(input: RDD[LabeledPoint], initialWeights: Vector): LogisticRegressionModel = {
+    run(input, initialWeights, true)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: `run(input, generateInitialWeights(input), userSuppliedWeights = true)`

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 12]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+   * defaults to the mllib implementation. If more than two classes
+   * or feature scaling is disabled, always uses mllib implementation.
+   * If using ml implementation, uses ml code to generate initial weights.
+   */
+  override def run(input: RDD[LabeledPoint]): LogisticRegressionModel = {
+    run(input, generateInitialWeights(input), false)
+  }
+
+  /**
+   * Run the algorithm with the configured parameters on an input RDD
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Replace `algorithm` by `Logistic Regression`, and add a new line.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 13]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+   * Uses user provided weights.
+   */
+  override def run(input: RDD[LabeledPoint], initialWeights: Vector): LogisticRegressionModel = {
+    run(input, initialWeights, true)
+  }
+
+  private def run(input: RDD[LabeledPoint], initialWeights: Vector, userSuppliedWeights: Boolean):
+      LogisticRegressionModel = {
+    // ml's Logisitic regression only supports binary classifcation currently.
+    if (numOfLinearPredictor == 1 && useFeatureScaling) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: You can remove `useFeatureScaling`, and pass it as setStandardization in ML implementation.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 14]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+        lr.setTol(optimizer.getConvergenceTol())
+        // Convert our input into a DataFrame
+        val sqlContext = new SQLContext(input.context)
+        import sqlContext.implicits._
+        val df = input.toDF()
+        // Determine if we should cache the DF
+        val handlePersistence = input.getStorageLevel == StorageLevel.NONE
+        if (handlePersistence) {
+          df.persist(StorageLevel.MEMORY_AND_DISK)
+        }
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Why do we need to do it? I through those check is already in ML code.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 15]
```diff
@@ -374,4 +383,82 @@ class LogisticRegressionWithLBFGS
...
+        // Train our model
+        val mlLogisticRegresionModel = lr.train(df)
+        // unpersist if we persisted
+        if (handlePersistence) {
+          df.unpersist()
+        }
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: ditto?

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 8 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: I don't think we know the number of features at that point.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 11 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: I'm assuming you meant `run(input, input, userSuppliedWeights = true)`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 8 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: You will know # of features by the size of coefficients set by setInitialModel. There is no ambiguity here since it's binary, and intercept has a separate variable.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 11 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: haha... yes.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 14 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: So the ML code checks on the DataFrame - which will never be cached. So we check on the user supplied input and if the user supplied input is not persisted we handle our own persistance but if the user supplied input is persisted then we don't.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 15 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: same

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 8 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Ah then there is no validation step we just assume if they set the initial model they set a valid initial model. Ok :)

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 14 above)
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: that makes sense.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 16]
```diff
@@ -247,8 +247,15 @@ class LogisticRegression @Since("1.2.0") (
   @Since("1.5.0")
   override def getThresholds: Array[Double] = super.getThresholds
 
-  override protected def train(dataset: DataFrame): LogisticRegressionModel = {
-    // Extract columns from data.  If dataset is persisted, do not persist oldDataset.
+  private var optInitialCoefficients: Option[Vector] = None
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Keep the reference to `InitialModel` here.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 17]
```diff
@@ -247,8 +247,15 @@ class LogisticRegression @Since("1.2.0") (
...
+
+  /** @group setParam */
+  private[spark] def setInitialModel(model: LogisticRegressionModel): this.type = {
+    this.optInitialCoefficients = Some(model.coefficients)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: You don't want to lose the information of intercept in model.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 18]
```diff
@@ -322,10 +329,11 @@ class LogisticRegression @Since("1.2.0") (
       new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
     }
 
-    val initialCoefficientsWithIntercept =
-      Vectors.zeros(if ($(fitIntercept)) numFeatures + 1 else numFeatures)
+    val numFeaturesWithIntercept = if ($(fitIntercept)) numFeatures + 1 else numFeatures
+    val initialCoefficientsWithIntercept = optInitialCoefficients.getOrElse(
+      Vectors.zeros(numFeaturesWithIntercept))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: here, 

```scala
val initialCoefficientsWithIntercept =	
   Vectors.zeros(if ($(fitIntercept)) numFeatures + 1 else numFeatures)

if (optInitialModel.isDefined && optInitialModel.get.coefficients.size == numFeatures) {
  val initialCoefficientsWithInterceptArray = initialCoefficientsWithIntercept.toArray
  optInitialModel.get.coefficients.foreachActive { case (index, value) =>
    initialCoefficientsWithInterceptArray(index) = value
  }
  if ($(fitIntercept) {
          initialCoefficientsWithInterceptArray(numFeatures) == optInitialModel.get.intercept
  } 
} else if ($(fitIntercept)) {
      /*
         For binary logistic regression, when we initialize the coefficients as zeros,
         it will converge faster if we initialize the intercept such that
         it follows the distribution of the labels.
         {{{
         P(0) = 1 / (1 + \exp(b)), and
         P(1) = \exp(b) / (1 + \exp(b))
         }}}, hence
         {{{
         b = \log{P(1) / P(0)} = \log{count_1 / count_0}
         }}}
       */
      initialCoefficientsWithIntercept.toArray(numFeatures)
        = math.log(histogram(1) / histogram(0))
    }
```

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: btw, may we want to log. `if (optInitialModel.isDefined && optInitialModel.get.coefficients.size != numFeatures)`, let's log it.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 19]
```diff
@@ -322,10 +329,25 @@ class LogisticRegression @Since("1.2.0") (
       new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
     }
 
+    val numFeaturesWithIntercept = if ($(fitIntercept)) numFeatures + 1 else numFeatures
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Is this used?

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 20]
```diff
@@ -343,8 +365,8 @@ class LogisticRegression @Since("1.2.0") (
         = math.log(histogram(1) / histogram(0))
     }
 
-    val states = optimizer.iterations(new CachedDiffFunction(costFun),
-      initialCoefficientsWithIntercept.toBreeze.toDenseVector)
+      val states = optimizer.iterations(new CachedDiffFunction(costFun),
+        initialCoefficientsWithIntercept.toBreeze.toDenseVector)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Wrong indentation. Remove two spaces.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 21]
```diff
@@ -374,4 +384,85 @@ class LogisticRegressionWithLBFGS
       new LogisticRegressionModel(weights, intercept, numFeatures, numOfLinearPredictor + 1)
     }
   }
+
+  /**
+   * Run Logistic Regression with the configured parameters on an input RDD
+   * of LabeledPoint entries starting from the initial weights provided.
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Removed `starting from the initial weights provided.` and add extra new line here for readability.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 22]
```diff
@@ -374,4 +384,85 @@ class LogisticRegressionWithLBFGS
...
+   * or feature scaling is disabled, always uses mllib implementation.
+   * If using ml implementation, uses ml code to generate initial weights.
+   */
+  override def run(input: RDD[LabeledPoint]): LogisticRegressionModel = {
+    run(input, generateInitialWeights(input), userSuppliedWeights = false)
+  }
+
+  /**
+   * Run Logistic Regression with the configured parameters on an input RDD
+   * of LabeledPoint entries starting from the initial weights provided.
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Add extra new line before `If a known updater is...`

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 23]
```diff
@@ -374,4 +384,85 @@ class LogisticRegressionWithLBFGS
...
+    // ml's Logisitic regression only supports binary classifcation currently.
+    if (numOfLinearPredictor == 1) {
+      def runWithMlLogisitcRegression(elasticNetParam: Double) = {
+        // Prepare the ml LogisticRegression based on our settings
+        val lr = new org.apache.spark.ml.classification.LogisticRegression()
+        lr.setRegParam(optimizer.getRegParam())
+        lr.setElasticNetParam(elasticNetParam)
+        lr.setStandardization(useFeatureScaling)
+        if (userSuppliedWeights) {
+          val initialWeightsWithIntercept = if (addIntercept) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: This is not used anymore.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 24]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
-             it follows the distribution of the labels.
-
-             {{{
-               P(0) = 1 / (1 + \exp(b)), and
-               P(1) = \exp(b) / (1 + \exp(b))
-             }}}, hence
-             {{{
-               b = \log{P(1) / P(0)} = \log{count_1 / count_0}
-             }}}
+        if (optInitialModel.isDefined && optInitialModel.get.coefficients != numFeatures) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: How can this compile? Should be `optInitialModel.get.coefficients.size != numFeatures`

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 25]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
+          val vec = optInitialModel.get.coefficients
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: `vec` is not used.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 26]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
+          logWarning(
+            s"Initial coefficients provided ${vec} did not match the expected size ${numFeatures}")
+        }
+
+        if (optInitialModel.isDefined && optInitialModel.get.coefficients == numFeatures) {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: ditto

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 27]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
+           * P(1) = \exp(b) / (1 + \exp(b))
+           * }}}, hence
+           * {{{
+           * b = \log{P(1) / P(0)} = \log{count_1 / count_0}
+           * }}}
            */
-          initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
-            histogram(1) / histogram(0))
+          initialCoefficientsWithIntercept.toArray(numFeatures)
+          = math.log(histogram(1) / histogram(0))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: add two spaces.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 28]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
+          }
+          if ($(fitIntercept)) {
+            initialCoefficientsWithInterceptArray(numFeatures) == optInitialModel.get.intercept
+          }
+        } else if ($(fitIntercept)) {
+          /**
+           * For binary logistic regression, when we initialize the coefficients as zeros,
+           * it will converge faster if we initialize the intercept such that
+           * it follows the distribution of the labels.
+
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: remove the extra line

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: I think u have to remove all the `*`. I think we decide to do comment like

```
/*
    Start the sentence.
 */

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 29]
```diff
@@ -335,31 +342,45 @@ class LogisticRegression @Since("1.2.0") (
...
 
         val states = optimizer.iterations(new CachedDiffFunction(costFun),
           initialCoefficientsWithIntercept.toBreeze.toDenseVector)
 
-        /*
-           Note that in Logistic Regression, the objective history (loss + regularization)
-           is log-likelihood which is invariance under feature standardization. As a result,
-           the objective history from optimizer is the same as the one in the original space.
+        /**
+         * Note that in Logistic Regression, the objective history (loss + regularization)
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: reverse the style change

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 30]
```diff
@@ -374,11 +395,11 @@ class LogisticRegression @Since("1.2.0") (
           throw new SparkException(msg)
         }
 
-        /*
-           The coefficients are trained in the scaled space; we're converting them back to
-           the original space.
-           Note that the intercept in scaled space and original space is the same;
-           as a result, no scaling is needed.
+        /**
+         * The coefficients are trained in the scaled space; we're converting them back to
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: ditto

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 31]
```diff
@@ -374,4 +384,82 @@ class LogisticRegressionWithLBFGS
...
+        }
+        lr.setFitIntercept(addIntercept)
+        lr.setMaxIter(optimizer.getNumIterations())
+        lr.setTol(optimizer.getConvergenceTol())
+        // Convert our input into a DataFrame
+        val sqlContext = new SQLContext(input.context)
+        import sqlContext.implicits._
+        val df = input.toDF()
+        // Determine if we should cache the DF
+        val handlePersistence = input.getStorageLevel == StorageLevel.NONE
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Will this cause double caching? Let's say input RDD is cached, so `handlePersistence` will be false. As a result, `df == StorageLevel.NONE` will be true in ml's LOR code, and this will cause caching twice.

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 32]
```diff
@@ -374,4 +384,82 @@ class LogisticRegressionWithLBFGS
...
+          df.persist(StorageLevel.MEMORY_AND_DISK)
+        }
+        // Train our model
+        val mlLogisticRegresionModel = lr.train(df)
+        // unpersist if we persisted
+        if (handlePersistence) {
+          df.unpersist()
+        }
+        // convert the model
+        val weights = mlLogisticRegresionModel.weights match {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]:
```scala
val weights = Vectors.dense(mlLogisticRegresionModel.coefficients.toArray)
```

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala [hunk 33]
```diff
@@ -374,4 +384,82 @@ class LogisticRegressionWithLBFGS
...
+          case x: DenseVector => x
+          case y: Vector => Vectors.dense(y.toArray)
+        }
+        createModel(weights, mlLogisticRegresionModel.intercept)
+      }
+      optimizer.getUpdater() match {
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: when `optimizer.getRegParam() == 0.0`, run the old version.

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: okay, this will make the test harder to write. I don't care this one now.

**Comment by dbtsai** [comment]
LGTM except some styling issues, and concern about caching twice. Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 25 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: its used on L348 in the log warning

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala (hunk 28 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: Ok, looking at the rest of the comments in the file & the style guide it seems to mostly have the `*` but I'll put them back in (it also break auto indent to not have them but thats an emacs bug)

File: mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala (hunk 31 above)
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala]: Good point, in a previous version of the code we passed handlePersistence down through to avoid this. I've updated it to do the same here.

**Comment by holdenk** [comment]
@dbtsai should have addressed the style concerns, let me know if anything else shows up :)

**Comment by dbtsai** [comment]
I'm going through the caching logic now. Will let you know soon. Thanks.

**Comment by holdenk** [comment]
Jenkins retest this please

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 34]
```diff
@@ -343,22 +355,36 @@ class LogisticRegression @Since("1.2.0") (
...
              }}}, hence
              {{{
-               b = \log{P(1) / P(0)} = \log{count_1 / count_0}
+             b = \log{P(1) / P(0)} = \log{count_1 / count_0}
              }}}
            */
-          initialCoefficientsWithIntercept.toArray(numFeatures) = math.log(
-            histogram(1) / histogram(0))
+          initialCoefficientsWithIntercept.toArray(numFeatures)
+            = math.log(histogram(1) / histogram(0))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: revert this change.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 35]
```diff
@@ -343,22 +355,36 @@ class LogisticRegression @Since("1.2.0") (
...
 
              {{{
-               P(0) = 1 / (1 + \exp(b)), and
-               P(1) = \exp(b) / (1 + \exp(b))
+             P(0) = 1 / (1 + \exp(b)), and
+             P(1) = \exp(b) / (1 + \exp(b))
              }}}, hence
              {{{
-               b = \log{P(1) / P(0)} = \log{count_1 / count_0}
+             b = \log{P(1) / P(0)} = \log{count_1 / count_0}
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: put two spaces back.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 36]
```diff
@@ -343,22 +355,36 @@ class LogisticRegression @Since("1.2.0") (
...
+        } else if ($(fitIntercept)) {
           /*
              For binary logistic regression, when we initialize the coefficients as zeros,
              it will converge faster if we initialize the intercept such that
              it follows the distribution of the labels.
 
              {{{
-               P(0) = 1 / (1 + \exp(b)), and
-               P(1) = \exp(b) / (1 + \exp(b))
+             P(0) = 1 / (1 + \exp(b)), and
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: put two spaces back.

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 37]
```diff
@@ -343,22 +355,36 @@ class LogisticRegression @Since("1.2.0") (
...
+             P(1) = \exp(b) / (1 + \exp(b))
```
Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: ditto

**Comment by dbtsai** [comment]
Thanks. Merged into master.
